doc/Requirements.md

```diff
@@ -0,0 +1,526 @@
+(Why: To allow introspection of the hardware.)
+
+A mode with synchronisation on client request MAY be supported.
+(Why: The reference implementation supports it.)
+
+# Rule based dynamic abstraction
+
+## Philosophy
+
+The PF offers parameters mapped on hardware. This is a good but weak abstraction.
+There is often a 1/1 relation between a parameter and the hardware it maps.
+
+A PF offers a mechanism to abstract the parameters to more high level concept.
+
+The goal is to hide numerous parameters and their dynamic value behind simple
+and human friendly api.
+
+It works by regrouping parameters with similar management and defining configurations
+for each "scenario". These "scenario" are then given a priority and a detection predicate.
+Configuration are applied when their associated "scenario" is detected.
+
+"Scenario" are detected through arbitrary criterion provided by the PF host.
+
+## Definition
+
+Configurations: set of values for different parameters.
+A configuration MUST not contain 2 values of the same parameters.
+For example, given a PF with 3 integer parameters A,B,C,
+a configuration can contain
+ - 1 value for (A), (B) or (C)
+ - 2 value for (A,B), (A,C) or (B,C)
+ - 3 value for (A,B,C)
+
+## Configuration
+
+A PF MUST offer configurations as described in the Definition chapter.
+(Note: rule based parameter engine does not manipulate directly values,
+      it applies configuration on the parameters.)
+(Why: This is what the reference implementation does.)
+
+Each configuration MUST be associated with a predicate that condition its eligibility.
+A configuration with a predicate that evaluates to `true` is called an "eligible configuration"
+(Why: This is what the reference implementation does.)
+
+The predicate SHOULD be a "selection criterion rule". See next chapter for a definition.
+(Why: The reference implementation uses a boolean expression based engine.)
+
+## Selection criterion
+
+A selection criterion MUST have one, and only one, state at a given time.
+
+A selection criterion MUST know at construction all possible states.
+(Why: To be able to validate:
+       - rules on start
+       - state changes)
+
+The selection criterion possible states MUST be specifiable by:
+ - Directly a state set (Input -> states == identity)
+   (Note: called exclusive criterion)
+   (Note: An empty set is not allowed as the criterion could not have a state.)
+   (Why: any criterion can be crated with this one.)
+
+The selection criterion possible states SHOULD be specifiable by:
+ - A combination of values
+   (Note: combination in the mathematical sense “ab” -> [“”, “a”, “b”, “ab”])
+   (Note: called inclusive criterion)
+   (Note: Un empty value set is allowed as it's combination -a set containing the empty set- would not be empty.
+          The empty set would be the only possible criteria state.)
+   (Why: The reference implementation supports it.)
+
```

Comment:
@dbrageul @scentelles : Is there a lower limit in the number of criterion/number of criterion states; does it depend on the kind or criterion (inclusive/exclusive) ? Or is it a MUST/SHOULD that there are no limits ?